server/data/data-library/src/test/java/org/apache/james/user/lib/AbstractUsersRepositoryTest.java

```diff
@@ -30,495 +30,519 @@
 import org.apache.james.core.MailAddress;
 import org.apache.james.core.Username;
 import org.apache.james.domainlist.api.mock.SimpleDomainList;
-import org.apache.james.lifecycle.api.LifecycleUtil;
 import org.apache.james.user.api.AlreadyExistInUsersRepositoryException;
 import org.apache.james.user.api.InvalidUsernameException;
 import org.apache.james.user.api.UsersRepositoryException;
 import org.apache.james.user.api.model.User;
 import org.junit.jupiter.api.Assumptions;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 
 
 public abstract class AbstractUsersRepositoryTest {
 
-    static final Domain DOMAIN = Domain.of("domain");
-
-    protected AbstractUsersRepository usersRepository;
-    protected SimpleDomainList domainList;
-
-    /**
-     * Create the repository to be tested.
-     *
-     * @return the user repository
-     * @throws Exception
-     */
-    protected abstract AbstractUsersRepository getUsersRepository() throws Exception;
-
-    Username user1;
-    Username user1CaseVariation;
-    Username user2;
-    Username user3;
-    Username admin;
-    Username adminCaseVariation;
-
-    protected void setUp() throws Exception {
-        domainList = new SimpleDomainList();
-        domainList.addDomain(DOMAIN);
-        this.usersRepository = getUsersRepository();
-        user1 = login("username");
-        user2 = login("username2");
-        user3 = login("username3");
-        user1CaseVariation = login("uSeRnaMe");
-        admin = login("admin");
-        adminCaseVariation = login("adMin");
-    }
-
-    protected void tearDown() throws Exception {
-        LifecycleUtil.dispose(this.usersRepository);
+    protected static class UserRepositoryExtension implements BeforeEachCallback, ParameterResolver {
```

Comment:
Contracts should not implement parameters resolution: they should be handled by implementations.

Reply:
Nothing relating to the contract here. I just put it inside a place that all implementations can reach to